apps/dcos_l4lb/src/dcos_l4lb_ipset_mgr.erl

```diff
@@ -0,0 +1,334 @@
+        {protocol, ?IPSET_PROTOCOL},
+        {setname, Name},
+        {data, [
+            {ip, [{Family, IP}]},
+            {port, Port},
+            {proto, Protocol}
+        ]}
+    ]).
+
+-spec(get_entries(pid(), Name :: string()) ->
+    {ok, [entry()]} | {error, atom() | non_neg_integer(), term()}).
+get_entries(Pid, Name) ->
+    Msg = [{protocol, ?IPSET_PROTOCOL}, {setname, Name}],
+    case request(Pid, list, [match, root, ack, request], Msg) of
+        {ok, Response} ->
+            {adt, ADT} = lists:keyfind(adt, 1, Response),
+            {ok, lists:map(fun ({data, Data}) ->
+                {ip, [{_Family, IP}]} = lists:keyfind(ip, 1, Data),
+                {port, Port} = lists:keyfind(port, 1, Data),
+                {proto, Protocol} = lists:keyfind(proto, 1, Data),
+                {Protocol, IP, Port}
+            end, ADT)};
+        {error, enoent, _Response} ->
+            {ok, []};
+        {error, Error, _Response} ->
+            {error, Error}
+    end.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec(request(pid(), Command :: atom(), Flags :: [atom()], Msg :: term()) ->
+    {ok, term()} | {error, atom() | non_neg_integer(), term()}).
+request(Pid, Command, Flags, Msg) ->
+    Args = [Pid, ?NETLINK_NETFILTER, ipset, Command, Flags, {inet, 0, 0, Msg}],
+    case apply(gen_netlink_client, request, Args) of
+        {ok, [{ipset, Command, _Flags, _Seq, _Pid, {inet, 0, 0, Response}}]} ->
+            {ok, Response};
+        {ok, Response} ->
+            {ok, Response};
+        {error, Code, Response} ->
+            Error = get_error(Code),
+            {error, Error, Response}
+    end.
+
+-spec(get_error(file:posix() | non_neg_integer()) ->
+    atom() | non_neg_integer()).
+get_error(Code) ->
+    % Link: git://git.netfilter.org/ipset.git
+    % File: include/libipset/linux_ip_set.h
+    Mapping = #{
+        4097 => ipset_err_protocol,
+        4098 => ipset_err_find_type,
+        4099 => ipset_err_max_sets,
+        4100 => ipset_err_busy,
+        4101 => ipset_err_exist_setname2,
+        4102 => ipset_err_type_mismatch,
+        4103 => ipset_err_exist,
+        4104 => ipset_err_invalid_cidr,
+        4105 => ipset_err_invalid_netmask,
+        4106 => ipset_err_invalid_family,
+        4107 => ipset_err_timeout,
+        4108 => ipset_err_referenced,
+        4109 => ipset_err_ipaddr_ipv4,
+        4110 => ipset_err_ipaddr_ipv6,
+        4111 => ipset_err_counter,
+        4112 => ipset_err_comment,
+        4113 => ipset_err_invalid_markmask,
+        4114 => ipset_err_skbinfo,
+        4352 => ipset_err_type_specific
+    },
+    maps:get(Code, Mapping, Code).
+
+%%%===================================================================
+%%% Cleanup function
+%%%===================================================================
+
+-spec(cleanup() -> ok).
+cleanup() ->
```

Comment:
Should we use the `maybe_`  version of add here so when cleanup is performed we do not add rules if ipv6 is not enabled for instance?

Reply:
This function is used only in tests and will always be used in test only. We must clean all the things that this module creates. It doesn't matter if ipv6 is disabled (btw, it's always enabled in tests) or if any command fails.